pallets/pallet-lottery/src/staking.rs

```diff
@@ -0,0 +1,495 @@
+            let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap_or_default();
+            let nonce: u128 = block_number ^ extrinsic_index as u128;
+            let randomness_output: sp_core::U256;
+            #[cfg(feature = "runtime-benchmarks")]
+            {
+                use rand::{Rng, SeedableRng};
+                let mut rng = rand::rngs::StdRng::seed_from_u64(nonce as u64);
+                randomness_output = rng.gen::<u128>().into();
+            }
+            #[cfg(not(feature = "runtime-benchmarks"))]
+            {
+                randomness_output = sp_core::U256::from_big_endian(
+                    T::RandomnessSource::random(&nonce.to_be_bytes()).0.as_ref(),
+                );
+            }
+            // NOTE: The following line introduces modulo bias, but since this is just a fallback it is accepted
+            let random_index: usize = randomness_output.low_u64() as usize % active_collators.len();
+            if let Some(random_collator) = active_collators.get(random_index) {
+                deposits.push((random_collator.clone(), remaining_deposit));
+                log::warn!(
+                    "Failed to select staking outputs. Staking {:?} randomly to {:?}",
+                    remaining_deposit,
+                    random_collator
+                );
+                remaining_deposit.set_zero();
+            }
+        }
+
+        if !remaining_deposit.is_zero() {
+            log::error!(
+                "We have {:?} unstaked balance left over after depositing",
+                remaining_deposit
+            );
+        }
+        if deposits.is_empty() {
+            log::error!("COULD NOT FIND ANY COLLATOR TO STAKE TO");
+        }
+        log::debug!("Deposits: {:?}", deposits);
+        deposits
+    }
+
+    #[named]
+    pub(crate) fn calculate_withdrawal_distribution(
+        withdrawal_amount: BalanceOf<T>,
+    ) -> Vec<T::AccountId> {
+        log::trace!(function_name!());
+        if withdrawal_amount.is_zero() {
+            return vec![];
+        }
+        let mut withdrawals = vec![];
+        let mut remaining_balance = withdrawal_amount;
+
+        // first concern: If there are inactive collators we are staked with, prefer these
+        let now = <frame_system::Pallet<T>>::block_number();
+        let info = pallet_parachain_staking::Pallet::<T>::round();
+        let active_collators: BTreeSet<_> =
+            pallet_parachain_staking::Pallet::<T>::selected_candidates()
+                .into_iter()
+                .filter(|collator| {
+                    // being selected is not enough, it must also be actively participating in block production
+                    // but we need to ensure it had a chance to produce blocks, so we only check this if we're at least 25% into a round
+                    if now < info.first + (info.length / 4u32).into() {
+                        return true;
```

Comment:
Does the logic rely on each collator producing at least one block ? The more the better ? Or that you will keep producing blocks ?

Reply:
only collators that have produced at least 1 block are considered active.
unless we're less than 25% into a round, then *all* selected collators are considered active.

Reply:
aha ok i get it now, but isn't 1/4 of a round too long . Can't we use the number of the candidates amount of blocks into the round ?

Reply:
people can randomly drop blocks sometimes, no need to unstake if they missed their first block or two for whatever reason

Reply:
discussion is obsolete, i changed it to check if this collator has earned any points in the last round instead